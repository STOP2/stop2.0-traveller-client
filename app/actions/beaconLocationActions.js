import Beacons from 'react-native-beacons-android';
import { DeviceEventEmitter } from 'react-native';

export const SET_BEACON_DATA = 'SET_BEACON_DATA';
export const SET_VEHICLE_BEACON_DATA = 'SET_VEHICLE_BEACON_DATA';
export const BEACON_ERROR = 'BEACON_ERROR';
export const VEHICLE_BEACON_ERROR = 'VEHICLE_BEACON_ERROR';
export const REQUEST_BEACON_DATA = 'REQUEST_BEACON_DATA';
export const REQUESTING_DATA = 'REQUESTING_DATA';

const FOREGROUND_SCAN_PERIOD = 1000;
const BACKGROUND_SCAN_PERIOD = 1000;

const beaconId = 'ebefd083-70a2-47c8-9837-e7b5634df524';
const vehicleBeaconId = 'f7826da6-4fa2-4e98-8024-bc5b71e0893e';

let beaconFound = false;
let vehicleBeaconsFound = false;

let tryingToFindBeacons = false;
let attempts = 0;

export const setBeaconData = function setBeaconData(beaconData) {
  return {
    type: SET_BEACON_DATA,
    beaconData,
    gettingBeaconData: false,
  };
};

export const setBusBeaconData = function setBusBeaconData(beaconData) {
  return {
    type: SET_VEHICLE_BEACON_DATA,
    beaconData,
    gettingVehicleBeaconData: false,
  };
};

export const requestBeaconData = function requestBeaconData() {
  return {
    type: REQUEST_BEACON_DATA,
    gettingBeaconData: true,
    gettingVehicleBeaconData: true
  };
};

export const beaconError = function beaconError(error) {
  return {
    type: BEACON_ERROR,
    beaconError: error,
    gettingBeaconData: false,
  };
};

export const vehicleBeaconError = function vehicleBeaconError(error) {
  return {
    type: VEHICLE_BEACON_ERROR,
    beaconError: error,
    gettingBeaconData: false,
  };
};

const getData = async function getData(dispatch) {
  Beacons.detectIBeacons();
  Beacons.setForegroundScanPeriod(FOREGROUND_SCAN_PERIOD);
  Beacons.setBackgroundScanPeriod(BACKGROUND_SCAN_PERIOD);

  try {
    await Beacons.startRangingBeaconsInRegion('STOPS', beaconId);
    await Beacons.startRangingBeaconsInRegion('BUSSES', vehicleBeaconId);
  } catch (error) {
    Beacons.stopRangingBeaconsInRegion('STOPS', beaconId);
    Beacons.stopRangingBeaconsInRegion('BUSSES', vehicleBeaconId);
    if (!beaconFound) dispatch(beaconError("Beacon didn't start ranging"));
    if (!vehicleBeaconsFound) dispatch(vehicleBeaconError("Beacon didn't start ranging"));
    tryingToFindBeacons = false;

    return;
  }

  // Print a log of the detected iBeacons (1 per 5 second)
  DeviceEventEmitter.addListener('beaconsDidRange', (data) => {
    if (attempts === 5) {
      Beacons.stopRangingBeaconsInRegion('STOPS', beaconId);
      Beacons.stopRangingBeaconsInRegion('BUSSES', vehicleBeaconId);
      if (!beaconFound) dispatch(beaconError('Beacon not found in 5 seconds'));
      if (!vehicleBeaconsFound) dispatch(vehicleBeaconError('Beacon not found in 5 seconds'));
      tryingToFindBeacons = false;

      return;
    }
    if (data.beacons.length > 0) {
      let closestBeaconIndex = 0;
      let closestBeaconDistance = 999999;

      let vehicleBeacons = [];

      for (let beaconIndex in data.beacons) {
        let beacon = data.beacons[beaconIndex];

        if (beacon.uuid === beaconId && beacon.distance < closestBeaconDistance) {
          closestBeaconIndex = beaconIndex;
          closestBeaconDistance = beacon.distance;
        }

        if (beacon.uuid === vehicleBeaconId) {
          vehicleBeacons.push({
            major: beacon.major,
            minor: beacon.minor,
          });
        }
      }

      let beaconData = {
        uuid: data.beacons[closestBeaconIndex].uuid,
        major: data.beacons[closestBeaconIndex].major,
        minor: data.beacons[closestBeaconIndex].minor,
      };

      if (beaconData.uuid === beaconId) {
        dispatch(setBeaconData(beaconData));
        Beacons.stopRangingBeaconsInRegion('STOPS', beaconId);
        beaconFound = true;
      }

      if (vehicleBeacons.length > 0) {
        dispatch(setBusBeaconData(vehicleBeacons));
        Beacons.stopRangingBeaconsInRegion('BUSSES', vehicleBeaconId);
        vehicleBeaconsFound = true;
      }
    }
    attempts++;
    if (beaconFound && vehicleBeaconsFound) tryingToFindBeacons = false;
  });
};

export const getBeaconData = function getBeaconData() {
  if (!tryingToFindBeacons) {
    tryingToFindBeacons = true;

    attempts = 0;
    beaconFound = false;
    vehicleBeaconsFound = false;

    return (dispatch) => {
      dispatch(requestBeaconData());
      getData(dispatch);
    };
  }

  return { type: REQUESTING_DATA };
};
