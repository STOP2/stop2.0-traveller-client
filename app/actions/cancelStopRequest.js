import config from '../config/config'

export const SEND_STOPREQUEST_FROM_CANCELLATION = 'SEND_STOPREQUEST_CANCELLATION'
export const STOPREQUEST_FROM_CANCELLED = 'STOPREQUEST_CANCELLED'
export const STOPREQUEST_FROM_CANCELLATION_ERROR ='STOPREQUEST_CANCELLATION_ERROR'

export const SEND_STOPREQUEST_DESTINATION_CANCELLATION = 'SEND_STOPREQUEST_CANCELLATION'
export const STOPREQUEST_DESTINATION_CANCELLED = 'STOPREQUEST_CANCELLED'
export const STOPREQUEST_DESTINATION_CANCELLATION_ERROR ='STOPREQUEST_CANCELLATION_ERROR'

const API_ENDPOINT = '/stoprequests/cancel'

export let requestStopRequestCancellation = function(fromVehicle)
{
    return {
        type: fromVehicle ? SEND_STOPREQUEST_FROM_CANCELLATION : SEND_STOPREQUEST_DESTINATION_CANCELLATION,
        stopRequestCancelled: false,
        isCancellingStopRequest: true
    }
}

export let receiveStopRequestCancellationConfirmation = function(fromVehicle)
{
    return {
        type: fromVehicle ? STOPREQUEST_FROM_CANCELLED : STOPREQUEST_DESTINATION_CANCELLED,
        stopRequestCancelled: true,
        isCancellingStopRequest: false
    }
}

export let stopRequestCancellationError = function(fromVehicle, error)
{
    return {
        type: fromVehicle ? STOPREQUEST_FROM_CANCELLATION_ERROR : STOPREQUEST_DESTINATION_CANCELLATION_ERROR,
        stopRequestCancelled: false,
        isCancellingStopRequest: false,
        stopRequestCancellationError: error
    }
}

export let cancelStopRequest = function(requestId, fromVehicle)
{
    return dispatch =>
    {
        dispatch(requestStopRequestCancellation())

        return fetch(config.API_URL + API_ENDPOINT + '?request_id=' + requestId, {
            method: 'POST',
            headers:
            {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }
        })
        .then(response => {
            if (response.ok) {
                dispatch(receiveStopRequestCancellationConfirmation(fromVehicle))

                // return false as there are no errors
                return false
            }
            else {
                dispatch(stopRequestCancellationError(fromVehicle))

                // return true as there is an error
                return true
            }
        })
        .catch(error => {
            dispatch(stopRequestCancellationError(error, fromVehicle))

            // return true as there is an error
            return true
        })
    }
}
