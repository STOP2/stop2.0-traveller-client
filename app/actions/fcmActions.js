export const SET_FCM_TOKEN = 'SET_FCM_TOKEN'

export let setFCMToken = function(token)
{
    return {
        type: SET_FCM_TOKEN,
        token: token
    }
}
