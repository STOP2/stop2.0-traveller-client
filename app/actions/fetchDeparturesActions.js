import config from '../config/config'

const USE_LOCATION_RADIUS = true
const LOCATION_RADIUS = 500

export const REQUEST_DEPARTURES = 'REQUEST_DEPARTURES'
export const RECEIVE_DEPARTURES = 'RECEIVE_DEPARTURES'
export const REQUEST_ERROR = 'REQUEST_ERROR'

const API_ENDPOINT = '/stops'
const API_ENDPOINT_WITH_BEACONS = '/stops/beacons'

export let requestDepartures = function(latitude, longitude)
{
    return {
        type: REQUEST_DEPARTURES,
        isFetching: true,
        latitude: latitude,
        longitude: longitude
    }
}

export let requestDeparturesWithBeacons = function(major, minor)
{
    return {
        type: REQUEST_DEPARTURES,
        isFetching: true,
        major: major,
        minor: minor
    }
}

export let receiveDepartures = function(json)
{
    return {
        type: RECEIVE_DEPARTURES,
        isFetching: false,
        isReady: true,
        error: false,
        departures: json.stops
    }
}

export let requestError = function()
{
    return {
        type: REQUEST_ERROR,
        error: true
    }
}

export let fetchDepartures = function(latitude, longitude, withBeacons)
{
    let radiusString = ''

    if (USE_LOCATION_RADIUS) radiusString = '&rad=' + LOCATION_RADIUS

    return dispatch =>
    {
        let fetchString

        if (withBeacons)
        {
            // here latitude =  major and longtitude = minor
            dispatch(requestDeparturesWithBeacons(latitude, longitude))
            fetchString = config.API_URL + API_ENDPOINT_WITH_BEACONS + '?major=' + latitude + '&minor=' + longitude + radiusString
            console.log('REQUESTING DEPARTURES WITH BEACONS')
            console.log(latitude)
            console.log(fetchString)
        }
        else
        {
            dispatch(requestDepartures(latitude, longitude))
            fetchString = config.API_URL + API_ENDPOINT + '?lat=' + latitude + '&lon=' + longitude + radiusString
        }


        return fetch(fetchString,
            {
                method: 'GET',
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                }
            })
            .then(response =>
              {
                if (response.ok)
                {
                    return response.json()
                }
                else
                {
                    dispatch(requestError())
                }
            })
            .then(json => dispatch(receiveDepartures(json)))
            .catch(error => dispatch(requestError(error)))
    }
}
