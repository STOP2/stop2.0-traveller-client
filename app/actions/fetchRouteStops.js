import config from '../config/config'

export const REQUEST_ROUTE_STOPS = 'REQUEST_ROUTE_STOPS'
export const RECEIVE_ROUTE_STOPS = 'RECEIVE_ROUTE_STOPS'
export const REQUEST_ROUTE_STOPS_ERROR = 'REQUEST_ROUTE_STOPS_ERROR'

export let requestRouteStops = function(tripId)
{
    return {
        type: REQUEST_ROUTE_STOPS,
        isFetching: true,
        tripId: tripId
    }
}

export let receiveRouteStops = function(json)
{
    return {
        type: RECEIVE_ROUTE_STOPS,
        isFetching: false,
        routeIsReady: true,
        error: false,
        stops: json.stops
    }
}

export let requestError = function(error)
{
    return {
        type: REQUEST_ROUTE_STOPS_ERROR,
        sentStoprequest: false,
        error: error
    }
}

export let fetchRouteStops = function(tripId)
{
    return dispatch =>
    {
        dispatch(requestRouteStops(tripId))

        return fetch(config.API_URL + '/routes?trip_id=' + tripId,
        {
            method: 'GET',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }
        })
      .then(response => response.json())
      .then(json => dispatch(receiveRouteStops(json)))
      .catch(error => dispatch(requestError(error)))
    }
}
