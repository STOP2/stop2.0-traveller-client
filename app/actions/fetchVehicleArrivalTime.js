import config from '../config/config'

export const REQUEST_VEHICLE_ARRIVAL_TIME = 'REQUEST_VEHICLE_ARRIVAL_TIME'
export const RECEIVE_ARRIVAL_TIME = 'RECEIVE_ARRIVAL_TIME'
export const REQUEST_ARRIVAL_TIME_ERROR = 'REQUEST_ARRIVAL_TIME_ERROR'

export let requestRouteStops = function(requestId)
{
    return {
        type: REQUEST_VEHICLE_ARRIVAL_TIME,
        isFetching: true,
        requestId: requestId
    }
}

export let receiveRouteStops = function(json)
{
    return {
        type: RECEIVE_ARRIVAL_TIME,
        isFetching: false,
        error: false,
        time: json.time
    }
}

export let requestError = function(error)
{
    return {
        type: REQUEST_ROUTE_STOPS_ERROR,
        sentStoprequest: false,
        error: error
    }
}

export let fetchRouteStops = function(tripId, stopId)
{
    return dispatch =>
    {
        dispatch(requestRouteStops(tripId, stopId))

        return fetch(config.API_URL + '/routes?trip_id=' + tripId + '&stop_id=' + stopId)
      .then(response => response.json())
      .then(json => dispatch(receiveRouteStops(json)))
      .catch(error => dispatch(requestError(error)))
    }
}
