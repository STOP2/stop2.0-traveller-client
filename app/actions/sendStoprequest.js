import config from '../config/config'
import strings from '../resources/translations'

export const SEND_STOPREQUEST = 'SEND_STOPREQUEST'
export const RECEIVE_CONFIRM = 'RECEIVE_CONFIRM'
export const REQUEST_ERROR = 'REQUEST_ERROR'

const API_ENDPOINT = '/stoprequests'

export let requestStoprequest = function(tripId, stopId, requestType)
{
    return {
        type: SEND_STOPREQUEST,
        sentStoprequest: false,
        tripId: tripId,
        stopId: stopId,
        requestType: requestType
    }
}

export let receiveConfirm = function()
{
    return {
        type: RECEIVE_CONFIRM,
        sentStoprequest: true
    }
}

export let requestError = function()
{
    alert(strings.stopRequestError)

    return {
        type: REQUEST_ERROR,
        error: true
    }
}

export let sendStoprequest = function(tripId, stopId, requestType)
{
    return dispatch =>
    {
        dispatch(requestStoprequest(tripId, stopId, requestType))

        let stopRequest = JSON.stringify({
            trip_id: tripId,
            stop_id: stopId,
            request_type: requestType
        })

        return fetch(config.API_URL + API_ENDPOINT, {
            method: 'POST',
            headers: {'Content-Type': 'application/json'},
            body: stopRequest
        })
            .then(response =>
            {
                if (response.ok)
                {
                    dispatch(receiveConfirm())

                    return response.json()
                }
                else
                {
                    dispatch(requestError())
                }
            })
            .catch(error => dispatch(requestError(error)))
    }
}
