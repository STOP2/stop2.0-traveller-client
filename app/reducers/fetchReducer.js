import { REQUEST_DEPARTURES, RECEIVE_DEPARTURES, REQUEST_ERROR } from '../actions/fetchDeparturesActions'
import { SEND_STOPREQUEST, RECEIVE_CONFIRM } from '../actions/sendStoprequest'

let initialState = {
    stops: [],
    isFetching: false,
    isReady: false,
    sentStoprequest: false,
    error: false,
    routeStops: {}
}

const fetchReducer = (state = initialState, action) =>
{
    switch (action.type)
    {
    case REQUEST_DEPARTURES:
        return Object.assign({}, state, {isFetching: action.isFetching})

    case RECEIVE_DEPARTURES:
        return Object.assign({}, state, {
            isFetching: action.isFetching,
            isReady: action.isReady,
            error: action.error,
            stops: action.departures
        }) // return only the first stop (temporarily)


    case SEND_STOPREQUEST:
        return Object.assign({}, state, {sentStoprequest: action.sentStoprequest})


    case RECEIVE_CONFIRM:
        return Object.assign({}, state, {sentStoprequest: action.sentStoprequest})

    case REQUEST_ERROR:
        return Object.assign({}, state, {error: action.error})

    default:
        return state
    }
}

export default fetchReducer
