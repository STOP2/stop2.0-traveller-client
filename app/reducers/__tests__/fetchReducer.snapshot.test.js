import fetchReducer, { initialState } from '../../reducers/fetchReducer'
import { requestDepartures, receiveDepartures, requestError } from '../../actions/fetchDeparturesActions'
import { requestStoprequest, receiveConfirm  } from '../../actions/sendStoprequest'
import { requestRouteStops, receiveRouteStops, requestError as requestRouteStopsError } from '../../actions/fetchRouteStops'

describe('fetch reducer', () =>
{
    it('returns the same state on an unhandled action', () =>
    {
        expect(fetchReducer(initialState, {type: '_NULL'})).toMatchSnapshot()
    })

    it('handles REQUEST_DEPARTURES action', () =>
    {
        expect(fetchReducer(initialState, requestDepartures(60.203978, 24.9633573))).toMatchSnapshot()
    })

    it('handles RECEIVE_DEPARTURES action', () =>
    {
        let departuresData = {"stops": [{"stop": {"schedule": [{"arrival": 7, "destination": "Rautatientori", "line": "55", "route_id": "HSL:1055", "trip_id": "HSL:1055_20161003_La_2_1412", "vehicle_type": 3}, {"arrival": 21, "destination": "Rautatientori", "line": "55", "route_id": "HSL:1055", "trip_id": "HSL:1055_20161003_La_2_1427", "vehicle_type": 3}], "stop_code": "3597", "stop_name": "A.I. Virtasen aukio"}}]}

        expect(fetchReducer(initialState, receiveDepartures(departuresData))).toMatchSnapshot()
    })

    it('handles REQUEST_ERROR action', () =>
    {
        expect(fetchReducer(initialState, requestError('error'))).toMatchSnapshot()
    })

    it('handles RECEIVE_DEPARTURES action', () =>
    {
        let departuresData = {"stops": [{"stop": {"schedule": [{"arrival": 7, "destination": "Rautatientori", "line": "55", "route_id": "HSL:1055", "trip_id": "HSL:1055_20161003_La_2_1412", "vehicle_type": 3}, {"arrival": 21, "destination": "Rautatientori", "line": "55", "route_id": "HSL:1055", "trip_id": "HSL:1055_20161003_La_2_1427", "vehicle_type": 3}], "stop_code": "3597", "stop_name": "A.I. Virtasen aukio"}}]}

        expect(fetchReducer(initialState, receiveDepartures(departuresData))).toMatchSnapshot()
    })

    it('handles RECEIVE_CONFIRM action', () =>
    {
        expect(fetchReducer(initialState, receiveConfirm())).toMatchSnapshot()
    })

    it('handles SEND_STOPREQUEST action', () =>
    {
        expect(fetchReducer(initialState, requestStoprequest(1, 2, 'stop'))).toMatchSnapshot()
    })

    it('handles REQUEST_ROUTE_STOPS action', () =>
    {
        expect(fetchReducer(initialState, requestRouteStops(1, 2))).toMatchSnapshot()
    })

    it('handles RECEIVE_ROUTE_STOPS action', () =>
    {
        expect(fetchReducer(initialState, receiveRouteStops({stops: []}))).toMatchSnapshot()
    })

    it('handles REQUEST_ROUTE_STOPS_ERROR action', () =>
    {
        expect(fetchReducer(initialState, requestRouteStopsError(true))).toMatchSnapshot()
    })
})