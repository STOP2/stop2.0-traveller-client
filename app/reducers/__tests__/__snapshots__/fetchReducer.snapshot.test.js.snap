exports[`fetch reducer handles RECEIVE_CONFIRM action 1`] = `
Object {
  "error": false,
  "isFetching": false,
  "isReady": false,
  "routeStops": Object {},
  "sentStoprequest": true,
  "stops": Array [],
}
`;

exports[`fetch reducer handles RECEIVE_DEPARTURES action 1`] = `
Object {
  "error": false,
  "isFetching": false,
  "isReady": true,
  "routeStops": Object {},
  "sentStoprequest": false,
  "stops": Array [
    Object {
      "stop": Object {
        "schedule": Array [
          Object {
            "arrival": 7,
            "destination": "Rautatientori",
            "line": "55",
            "route_id": "HSL:1055",
            "trip_id": "HSL:1055_20161003_La_2_1412",
            "vehicle_type": 3,
          },
          Object {
            "arrival": 21,
            "destination": "Rautatientori",
            "line": "55",
            "route_id": "HSL:1055",
            "trip_id": "HSL:1055_20161003_La_2_1427",
            "vehicle_type": 3,
          },
        ],
        "stop_code": "3597",
        "stop_name": "A.I. Virtasen aukio",
      },
    },
  ],
}
`;

exports[`fetch reducer handles RECEIVE_DEPARTURES action 2`] = `
Object {
  "error": false,
  "isFetching": false,
  "isReady": true,
  "routeStops": Object {},
  "sentStoprequest": false,
  "stops": Array [
    Object {
      "stop": Object {
        "schedule": Array [
          Object {
            "arrival": 7,
            "destination": "Rautatientori",
            "line": "55",
            "route_id": "HSL:1055",
            "trip_id": "HSL:1055_20161003_La_2_1412",
            "vehicle_type": 3,
          },
          Object {
            "arrival": 21,
            "destination": "Rautatientori",
            "line": "55",
            "route_id": "HSL:1055",
            "trip_id": "HSL:1055_20161003_La_2_1427",
            "vehicle_type": 3,
          },
        ],
        "stop_code": "3597",
        "stop_name": "A.I. Virtasen aukio",
      },
    },
  ],
}
`;

exports[`fetch reducer handles REQUEST_DEPARTURES action 1`] = `
Object {
  "error": false,
  "isFetching": true,
  "isReady": false,
  "routeStops": Object {},
  "sentStoprequest": false,
  "stops": Array [],
}
`;

exports[`fetch reducer handles REQUEST_ERROR action 1`] = `
Object {
  "error": true,
  "isFetching": false,
  "isReady": false,
  "routeStops": Object {},
  "sentStoprequest": false,
  "stops": Array [],
}
`;

exports[`fetch reducer handles SEND_STOPREQUEST action 1`] = `
Object {
  "error": false,
  "isFetching": false,
  "isReady": false,
  "routeStops": Object {},
  "sentStoprequest": false,
  "stops": Array [],
}
`;

exports[`fetch reducer returns the same state on an unhandled action 1`] = `
Object {
  "error": false,
  "isFetching": false,
  "isReady": false,
  "routeStops": Object {},
  "sentStoprequest": false,
  "stops": Array [],
}
`;
